### ISSUE TITLE
How to 'accuracy' is calculated in different cases in sequence learning using RNN?

### ISSUE THREAD

**Issue by fluency03** [fchollet/keras#2275]
I am wondering how the accuracy in training and validation is calculated in sequence learning using RNN? Let's say I have two cases just examples for showing my problems) as following:

(1). 

    X_train       |    y_train 
    ABCDEFGHIJ    |       K 
    BCDEFGHIJK    |       L
    CDEFGHIJKL    |       M
    DEFGHIJKLM    |       N
    EFGHIJKLMN    |       O
    ......

(2). 

    X_train       |    y_train 
    ABCDEFGHIJ    |    BCDEFGHIJK
    BCDEFGHIJK    |    CDEFGHIJKL
    CDEFGHIJKL    |    DEFGHIJKLM
    DEFGHIJKLM    |    EFGHIJKLMN
    EFGHIJKLMN    |    FGHIJKLMNO
    ......

And this is how I understand the accuracy. In general, accuracy=#correct_output_y / #total_y 

So, in the first case, as long as the outputted y is equal to the targeted y, then it is correct, where there is only one char to be considered. 

However, in the second case, only when all chars in one outputted y are equal to all chars in targeted y, the output will be considered as correct one. 

Find the accuracy calculation [here](https://github.com/fchollet/keras/blob/master/keras/metrics.py):

    def categorical_accuracy(y_true, y_pred):
        return K.mean(K.equal(K.argmax(y_true, axis=-1),
                      K.argmax(y_pred, axis=-1)))

But when calculating the loss in the second case, even though most of the chars in one outputted y is correct, which leads to a low loss value, the output y is still considered as wrong when calculating the accuracy. 

Am I correct?

**Comment by braingineer** [comment]
The categorical accuracy is doing pairwise matches for item-wise sequence classifications.   So if you had


y_true                 |   y_pred
-----------------------|----------------------
the fish swam   | the bird swam

then this would be 0.66 accuracy.   I'm not sure what you mean by "the output y is still considered wrong".   It's not a binary metric. 

Be careful if you use masking, though. #2260 .

**Comment by fluency03** [comment]
@braingineer Oh. It seems like my understanding is wrong. 

How I understand is like this:

| y_true | y_pred |
| --- | --- |
| aaa | aaa |
| bbb | bbb |
| ccc | ccc |
| ddd | ddc |
| eee | eee |

There are 5 data points, 4 of them are correct, then the accuracy is 4/5=80% .

Thanks.

**Comment by braingineer** [comment]
Nope.  It does pairwise comparisons. 

`K.argmax(y_pred, axis=-1)` is over the prediction dimension. 

Your dimensions should be:

`( batch, sequence length, vocabulary size) `

The last dimension classification layer on top of your RNN should be of output size `vocabulary size`.   

Additionally, your y_truth should be of shape `batch, sequence length, vocabulary size` as well, but the vocabulary size should be a 1-hot.  there is a function to aid in this: `keras.utils.np_utils.to_categorical` which takes as input a vector of integers and outputs a matrix of 1-hots. 

For example, I have encoded as such:
```python


    def serve_sentence(self, data):
        for data_i in np.random.choice(len(data), len(data), replace=False):
            in_X = np.zeros(self.max_sequence_len)
            out_Y = np.zeros(self.max_sequence_len, dtype=np.int32)
            bigram_data = zip(data[data_i][0:-1], data[data_i][1:])
            for datum_j,(datum_in, datum_out) in enumerate(bigram_data):
                in_X[datum_j] = datum_in
                out_Y[datum_j] = datum_out
            yield in_X, out_Y

    def serve_batch(self, data):
        dataiter = self.serve_sentence(data)
        V = self.vocab_size
        S = self.max_sequence_len
        B = self.batch_size

        while dataiter:
            in_X = np.zeros((B, S), dtype=np.int32)
            out_Y = np.zeros((B, S, V), dtype=np.int32)
            next_batch = list(itertools.islice(dataiter, 0, self.batch_size))
            if len(next_batch) < self.batch_size:
                raise StopIteration
            for d_i, (d_X, d_Y) in enumerate(next_batch):
                in_X[d_i] = d_X
                out_Y[d_i] = to_categorical(d_Y, V)
                
            yield in_X, out_Y
```

**Comment by braingineer** [comment]
Sorry, I should also have mentioned that it matters if your RNN is serving the last output or the sequence output.   If `return_sequences=True` then it behaves as I said.  If not, then it only returns the last output.
